### ISSUE TITLE
Feature request: display minor mode state in command annotation

### ISSUE THREAD

**Issue by jumper047** [minad/marginalia#155]
Recently switched to vertico from ivy, and missed one convenient feature - ivy highlighted minor-mode commands when corresponding minor modes were activated. I found myself relying on it often (I use doom mode line which hides many minor mode lighters, and ivy highlighing was convenient way to understand is the command will turn on or off something). Can you add similiar feature to marginalia (or what do you think about PR containing this feature)?

**Comment by minad** [comment]
Highlighting candidates is out of scope of Marginalia. But you can easily implement candidate transformations in Vertico as part of your configuration, see https://github.com/minad/vertico/wiki#candidate-display-transformations-custom-candidate-highlighting.

**Comment by jumper047** [comment]
Hmm, thank you for the link, I'll check it! But I thought this is exactly what Marginalia supposed to do - display additional information about candidate. Ivy made it using highlighting, Marginalia can do it with additional symbol before docstring (I suppose, though, Marginalia still can do highlighting with faces), like "[On] Some minor mode ...".

**Comment by minad** [comment]
No, Marginalia does not aim to modify the candidate strings itself. We could implement candidate highlighting via the `affixation-function` as part of Marginalia, but I prefer if Marginalia is more limited in scope and if candidates are passed through unmodified as created by the backend. 

Furthermore I consider highlighting of candidates as not important. I've only found two use cases for it: highlighting directories and highlighting modes. These two use cases are simply not worth adding complexity here, given that you can add candidate transformations to Vertico.
Note that Consult also offers the command `consult-minor-mode`.

**Comment by minad** [comment]
> ...Marginalia can do it with additional symbol before docstring (I suppose, though, Marginalia still can do highlighting with faces), like "[On] Some minor mode ...".

This is something we could add. I will give it a try.

**Comment by jumper047** [comment]
Thank you! Personally I'll be fine with the highlighter function in my config, but I still sure this little addition will be useful for other Marginalia users

**Comment by hmelman** [comment]
Wasn't really sure I followed this.  `consult-minor-mode-menu` with marginalia does show whether the mode is on or off, just not by highlighting the candidate itself.  But if you mean indicating if a minor-mode is on/off in something like `M-x`, perhaps `marginalia-annotate-command` could be modified to preface the doc string with On/Off if the command is a minor-mode command (in your config or perhaps marginalia).

**Comment by minad** [comment]
I implemented this in https://github.com/minad/marginalia/tree/mode-state, but I don't really like the result, since only a minority of the commands are modes. This is also why I don't like the highlighting idea. I could format it a bit differently, such that the On/Off doesn't use its own column, but the result is still not pretty. You can copy this annotator to your user configuration if you like.

**Comment by oantolin** [comment]
Yeah, I also thought the `[On]` would look ugly if it didn't get it's own column, but that if it _did_, then it would be a waste of space, since the vast majority of commands are not minor mode toggles. This of course does not apply to the `consult-minor-mode-menu` command @minad mentioned since it _only_ shows minor modes and it does annotate candidate with `On` or `Off`. That's the command I'd recommend to check the status of minor modes, @jumper047.

**Comment by minad** [comment]
@oantolin 

> Yeah, I also thought the [On] would look ugly if it didn't get its own column, but that if it did, then it would be a waste of space, since the vast majority of commands are not minor mode toggle.

Yes, exactly. I also tried to put the On/Off directly behind the key binding, but this is also ugly. Not every command is bound to a key of course, but I consider the key bindings far more important. I often use M-x as a reminder of some rarely used bindings if I remember parts of the command name.

**Comment by jumper047** [comment]
FYI: 
I realized that marginalia is not the only package which provides metadata for completion:)
Just added a method to the all-the-icons-completion:
```
(cl-defmethod all-the-icons-completion-get-icon (cand (_cat (eql command)))
  "Return the icon for the candidate CAND of completion category command."

  (let* ((mode-p (string-suffix-p "-mode" cand))
         (mode-symbol (intern cand))
         (mode-enabled (and (boundp mode-symbol) (symbol-value mode-symbol)))
         (icon-name (if mode-p "cogs" "cog"))
         (icon-face (when (and mode-p mode-enabled) 'all-the-icons-blue)))
    (concat (all-the-icons-faicon icon-name :height 0.95 :v-adjust -0.05 :face icon-face) " ")))
```

![image](https://user-images.githubusercontent.com/1407838/224261390-cb72ff6f-ed02-4cec-87d2-ce44bca070d6.png)

**Comment by minad** [comment]
Yes, `all-the-icons-completion` was modeled after Marginalia to specifically provide icons which are out of scope of Marginalia. You should find the origins of this package on the Marginalia issue tracker. :)

**Comment by panurg** [comment]
Minor mode state annotation seems to be a popular request (see #75 ). I just want to share my approach based on @minad solution and some other Marginalia parts:

```lisp
  (defun marginalia--mode-state (mode)
    "Return MODE state string."
    (if (and (boundp mode) (symbol-value mode))
        #(" [On]" 1 5 (face marginalia-on))
      #(" [Off]" 1 6 (face marginalia-off))))
  (defun marginalia--annotate-minor-mode-command (orig cand)
    "Annotate minor-mode command CAND with mode state."
    (concat
     (when-let* ((sym (intern-soft cand))
                 (mode (if (and sym (boundp sym))
                           sym
                         (lookup-minor-mode-from-indicator cand))))
       (marginalia--mode-state mode))
     (funcall orig cand)))
  (advice-add #'marginalia-annotate-command
              :around #'marginalia--annotate-minor-mode-command)
```
State annotation format is similar to the bindings annotation. Functions ending on "-mode", autoloaded modes and major modes are not annotated. Looks rather pretty for me:
![image](https://user-images.githubusercontent.com/3534692/226739906-424e81a4-0a4c-4dea-a53d-cb951c23e3fa.png)
Maybe it could be useful for someone, but keep in mind that I am not familiar with Lisp, so this code could have some potential issues, e.g. poor performance.

**Comment by jumper047** [comment]
@minad sorry for tagging but what do you think about solution posted by @panurg? imho it looks awesome, may be this approach is good enough to be included into upstream?
